$MirOS: ports/devel/cvs/patches/patch-src_patch_c,v 1.1 2007/03/10 23:53:50 tg Exp $
--- src/patch.c.orig	Fri Sep 23 02:02:42 2005
+++ src/patch.c	Thu Sep 13 14:04:27 2007
@@ -20,6 +20,8 @@
 #include "cvs.h"
 #include "getline.h"
 
+__RCSID("$MirOS: src/gnu/usr.bin/cvs/src/patch.c,v 1.2 2007/02/16 19:41:59 tg Exp $");
+
 static RETSIGTYPE patch_cleanup (int);
 static Dtype patch_dirproc (void *callerdat, const char *dir,
                             const char *repos, const char *update_dir,
@@ -46,13 +48,13 @@ static int unidiff = 0;
 
 static const char *const patch_usage[] =
 {
-    "Usage: %s %s [-flR] [-c|-u] [-s|-t] [-V %%d] [-k kopt]\n",
+    "Usage: %s %s [-flR] [-c|-u[p]] [-s|-t] [-V %%d] [-k kopt]\n",
     "    -r rev|-D date [-r rev2 | -D date2] modules...\n",
     "\t-f\tForce a head revision match if tag/date not found.\n",
     "\t-l\tLocal directory only, not recursive\n",
     "\t-R\tProcess directories recursively.\n",
     "\t-c\tContext diffs (default)\n",
-    "\t-u\tUnidiff format.\n",
+    "\t-u\tUnidiff format (-p like diff).\n",
     "\t-s\tShort patch - one liner per file.\n",
     "\t-t\tTop two diffs - last change made to the file.\n",
     "\t-V vers\tUse RCS Version \"vers\" for keyword expansion.\n",
@@ -78,7 +80,7 @@ patch (int argc, char **argv)
 	usage (patch_usage);
 
     optind = 0;
-    while ((c = getopt (argc, argv, "+V:k:cuftsQqlRD:r:")) != -1)
+    while ((c = getopt (argc, argv, "+V:k:cupftsQqlRD:r:")) != -1)
     {
 	switch (c)
 	{
@@ -149,11 +151,14 @@ patch (int argc, char **argv)
 		       "the -V option is obsolete and should not be used");
 		break;
 	    case 'u':
-		unidiff = 1;		/* Unidiff */
+		unidiff |= 1;		/* Unidiff */
 		break;
 	    case 'c':			/* Context diff */
-		unidiff = 0;
+		unidiff &= ~1;
 		break;
+	    case 'p':
+		unidiff |= 2;		/* Unidiff context */
+		break;
 	    case '?':
 	    default:
 		usage (patch_usage);
@@ -167,6 +172,8 @@ patch (int argc, char **argv)
     if (argc < 1)
 	usage (patch_usage);
 
+    if (!(unidiff & 1))
+	unidiff = 0;
     if (toptwo_diffs && patch_short)
 	error (1, 0, "-t and -s options are mutually exclusive");
     if (toptwo_diffs && (date1 != NULL || date2 != NULL ||
@@ -202,6 +209,8 @@ patch (int argc, char **argv)
 	    send_arg("-s");
 	if (unidiff)
 	    send_arg("-u");
+	if (unidiff & 2)
+	    send_arg("-p");
 
 	if (rev1)
 	    option_with_arg ("-r", rev1);
@@ -270,6 +279,7 @@ patch_proc (int argc, char **argv, char 
     int which;
     char *repository;
     char *where;
+    char *cp;
 
     TRACE ( TRACE_FUNCTION, "patch_proc ( %s, %s, %s, %d, %d, %s, %s )",
 	    xwhere ? xwhere : "(null)",
@@ -292,7 +302,6 @@ patch_proc (int argc, char **argv, char 
     /* if mfile isn't null, we need to set up to do only part of the module */
     if (mfile != NULL)
     {
-	char *cp;
 	char *path;
 
 	/* if the portion of the module is a path, put the dir part on repos */
@@ -342,14 +351,30 @@ patch_proc (int argc, char **argv, char 
 
     if (rev1 != NULL && !rev1_validated)
     {
-	tag_check_valid (rev1, argc - 1, argv + 1, local_specified, 0,
-			 repository, false);
+	if ((cp = strchr(rev1, ':')) != NULL)
+	{
+	    *cp++ = '\0';
+	    date1 = Make_Date (cp);
+	    if (*rev1 == '\0')
+		rev1 = NULL;
+	}
+	if (rev1)
+	    tag_check_valid (rev1, argc - 1, argv + 1, local_specified, 0,
+			     repository, false);
 	rev1_validated = 1;
     }
     if (rev2 != NULL && !rev2_validated)
     {
-	tag_check_valid (rev2, argc - 1, argv + 1, local_specified, 0,
-			 repository, false);
+	if ((cp = strchr(rev2, ':')) != NULL)
+	{
+	    *cp++ = '\0';
+	    date2 = Make_Date (cp);
+	    if (*rev2 == '\0')
+		rev2 = NULL;
+	}
+	if (rev2)
+	    tag_check_valid (rev2, argc - 1, argv + 1, local_specified, 0,
+			     repository, false);
 	rev2_validated = 1;
     }
 
@@ -571,6 +596,7 @@ patch_fileproc (void *callerdat, struct 
 
     if (unidiff) run_add_arg_p (&dargc, &darg_allocated, &dargv, "-u");
     else run_add_arg_p (&dargc, &darg_allocated, &dargv, "-c");
+    if (unidiff & 2) run_add_arg_p (&dargc, &darg_allocated, &dargv, "-p");
     switch (diff_exec (tmpfile1, tmpfile2, NULL, NULL, dargc, dargv,
 		       tmpfile3))
     {
@@ -671,7 +697,10 @@ failed to read diff file header %s for %
 	       program.  */
 	    if (unidiff)
 	    {
-		cvs_output ("diff -u ", 0);
+		if (unidiff & 2)
+		    cvs_output ("diff -up ", 0);
+		else
+		    cvs_output ("diff -u ", 0);
 		cvs_output (file1, 0);
 		cvs_output (" ", 1);
 		cvs_output (file2, 0);
